File: 46.permutations.java
/*
 * @lc app=leetcode id=46 lang=java
 *
 * [46] Permutations
 */

// @lc code=start
class Solution {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> permute(int[] nums) {
        if (nums != null && nums.length > 0) {
            backtrack(1, nums, new LinkedList<Integer>() {{add(nums[0]);}} );
        }
        return res;
    }

    private void backtrack(int idx, int[] nums, List<Integer> tmpList) {
        if (idx == nums.length) {
            res.add(new ArrayList<>(tmpList));
            return;
        }
        for (int i = 0; i <= tmpList.size(); ++i) {
            tmpList.add(i, nums[idx]);
            backtrack(idx+1, nums, tmpList);
            tmpList.remove(i);
        }
    }



}
// @lc code=end

File: go/46.permutations.go
/*
 * @lc app=leetcode id=46 lang=golang
 *
 * [46] Permutations
 */

// @lc code=start
func permute(nums []int) [][]int {
	res := [][]int{}
	if len(nums) == 0 {
		return res
	}
	bitmap := make([]bool, len(nums))
	var backtrack func([]int)
	backtrack = func(per []int) {
		if len(per) == len(nums) {
			tmp := make([]int, len(nums))
			copy(tmp, per)
			res = append(res, tmp)
			return
		}
		for i := 0; i < len(nums); i++ {
			if bitmap[i] {
				continue
			}
			bitmap[i] = true
			per = append(per, nums[i])
			backtrack(per)
			per = per[:len(per)-1]
			bitmap[i] = false
		}
	}
	backtrack([]int{})
	return res
}
// @lc code=end

